homeassistant/components/unifi_access/coordinator.py

```diff
@@ -0,0 +1,101 @@
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class UniFiAccessDoorCoordinator(DataUpdateCoordinator[dict[str, uiaccessclient.Door]]):
+    """Handles refreshing door data."""
+
+    def __init__(
+        self, hass: HomeAssistant, api_client: uiaccessclient.ApiClient
+    ) -> None:
+        """Initialize the door coordinator class."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="UniFi Access door",
+            always_update=False,
+        )
+
+        self.task: Task[None] | None = None
+        self.configuration = api_client.configuration
+        self.space_api = uiaccessclient.SpaceApi(api_client)
+
+    async def _async_setup(self) -> None:
+        self.task = self.hass.async_create_task(
+            self.receive_updated_data(), eager_start=True
+        )
+
+    async def async_shutdown(self) -> None:
+        """Cancel any scheduled call, and ignore new runs."""
+        await super().async_shutdown()
+
+        if self.task is not None:
+            self.task.cancel()
+            with suppress(CancelledError):
+                await self.task
+
+    async def _async_update_data(self) -> dict[str, uiaccessclient.Door]:
+        return await self.hass.async_add_executor_job(self._update_data)
+
+    async def receive_updated_data(self) -> None:
+        """Start websocket receiver for updated data from UniFi Access."""
+        _LOGGER.debug(
+            "Starting UniFi Access websocket with %s", self.configuration.host
+        )
+        try:
+            async with (
+                aiohttp.ClientSession(
+                    base_url=self.configuration.host.rstrip("/") + "/",
+                    headers={
+                        "Authorization": f"Bearer {self.configuration.access_token}"
+                    },
+                ) as session,
+                session.ws_connect(
+                    "devices/notifications", verify_ssl=self.configuration.verify_ssl
+                ) as socket,
+            ):
+                # WebSocket API is poorly documented so we will just use the REST API whenever we get
+                # an update to fetch all the relevant data.
+                async for message in socket:
+                    json = message.json()
+                    if (
+                        type(json) is dict
+                        and json.get("event") == "access.data.v2.device.update"
+                    ):
+                        _LOGGER.debug(
+                            "Received update from UniFi Access: %s", json.get("event")
+                        )
+                        self.async_set_updated_data(
+                            await self.hass.async_add_executor_job(self._update_data)
+                        )
+        except Exception as exc:
+            _LOGGER.error("Error in UniFi Access websocket receiver: %s", exc)
+            raise
```

Comment:
This should be abstracted away in the library

Reply:
Agreed, will add it to library. This is here as library as of now is just a generated OpenAPI client; and anything outside the scope of OpenAPI like websocket I've not yet implemented in the library. I'll add some extra custom endpoints for websocket to the library on top of the OpenAPI generated code to resolve this.

Reply:
My 2 cents, don't use an OpenAPI generated library. Yes it works quick and easy, but it will end you up with an integration that has to make up for the fact that the library is generated.

Reply:
This is now moved to library.